### ISSUE TITLE
Conda packaging

### ISSUE THREAD

**Issue by lohedges** [michellab/Sire#261]
It turns out that creating a Conda package is more tricky than anticipated. One of the main problems is that we currently don't pin versions for the additional Conda dependencies installed within the [build_sire.py](https://github.com/michellab/Sire/blob/devel/build/build_sire.py) script. This means that the state of the Conda environment is non-reproducible. I get a different environment on different Linux machines so the Conda package files that I create during the Azure pipeline build aren't guaranteed to work in my local Conda environment. (There are also, less surprisingly, differences between the Linux and macOS environment too.)

Since we compile Sire against specific versions of external librabries, e.g. boost, tbb, gsl, we will need to match these *exact* versions in the Conda Forge feedstock. In contrast, if we were to actually build Sire on the Conda Forge platform, then we wouldn't need to specify versions for the dependencies since these would be determined automatically. (We'd obviously specify versions if only particular versions of the dependencies worked.) Also, there could potentially be different Sire Conda packages built against different versions of the dependencies.

Another solution is to use a [script](https://github.com/michellab/Sire/blob/feature-conda/docker/sire-conda-devel/update_feedstock.sh) that I have written which runs as part of the pipeline build. This queries the SIre Conda environment to get the versions of the dependencies, then updates the feedstock accordingly. It also gets the latest Sire version so that the Conda package version can be incremented and also sets the sha256 hashes for the Conda package archives. However, since the dependency versions aren't necessarily consistent between the Linux and macOS build, we'll need to manage two sets of the run-time requirements in the feedstock meta.yaml. (These can be labeled with `# [linux]` and `# [osx]`.)

In summary, possible options are:

1) Pin versions of external dependencies in [build_sire.py](https://github.com/michellab/Sire/blob/devel/build/build_sire.py), match these in the Conda Forge feedstock (could be automatic), and update periodically as necessary.

2) Don't pin versions and automatically generate the feedstock run-time requirements during the Azure pipeline. (Already have code to do this.) Will probably have to manage requirements for both Linux and macOS (and Windows...) Conda package. We will need to dynamically update the package hashes and version number anyway. 

I'm not sure what the best solution is and welcome suggestions.

**Comment by chryswoods** [comment]
I think pinning versions is better. We can then review and update versions under our control, thereby minimising support headache of not knowing what versions a random user’s conda is using (or dealing with random test failures or weirdness in the nightly builds).

I really didn’t want to pin versions, but conda is too painful to use otherwise. I don’t see any other way to solve this on our own, and should pin until we can join with, eg bioexcel/bioconda/molssi to jointly work out a better strategy. At the end of the day, conda versioning and packaging is a mess and we shouldn’t spend too much time trying to fix or work around it :-(

**Comment by lohedges** [comment]
This problem is partly mitigated by prepending the conda-forge channel at the start of the build_sire.py script. That way, as long as there are no package updates in the mean time, we end up with the same set of packages in the Sire miniconda as in a local Python 3.7 conda environment. Another benefit is that the conda-forge channel uses openblas as its default so we avoid the numpy issues with MKL blas.

I think I'll update the build_sire.py script so that it reads in a file (perhaps JSON/YAML) that lists the conda/pip build and runtime dependencies (like a requirements.txt). This will make it easier to manage versions and to figure out what's needed in the conda-forge feedstock. (Although it's easy enough to use 'conda env' to do this.)

Just a note that I am still seeing [this](https://github.com/michellab/Sire/issues/253) issue during the wrapper compilation stage. @ppxasjsm had the same problem on her Linux box.

**Comment by lohedges** [comment]
Closing since we now have a working Conda package available [here](https://anaconda.org/michellab/sire).
